### PR TITLE
Add 'xresources' theme / set/handle DPI

### PR DESCRIPTION
**Pull request by actionless**
It follows xrdb colorscheme and have a hack for HiDPI support:
http://i.imgur.com/jmIV5Wk.png
UPD: http://i.imgur.com/d8GwbQF.png

### CODE REVIEW

File: themes/xresources/theme.lua [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+    cr:line_to(0, size)
+    cr:line_to(0, 0)
+    cr:stroke()
+    return img
+end
+theme.taglist_squares_unsel = taglist_squares_unsel()
+
+
+return theme
+-- vim: filetype=lua:expandtab:shiftwidth=4:tabstop=8:softtabstop=4:textwidth=80
```
Comment (blueyed) [inline themes/xresources/theme.lua]: Minor note: maybe we should switch to providing an .editorconfig file instead?

**Comment by blueyed** [comment]
Sweet!
It nicely picks up my Solarized theme! :)

But you should provide a `.in` file instead, and use `@AWESOME_THEMES_PATH@` and `@AWESOME_ICON_PATH@` therein.

**Comment by actionless** [comment]
Fixed.

File: themes/xresources/theme.lua.in [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+theme.titlebar_floating_button_focus_active  = "@AWESOME_THEMES_PATH@/default/titlebar/floating_focus_active.png"
+
+theme.titlebar_maximized_button_normal_inactive = "@AWESOME_THEMES_PATH@/default/titlebar/maximized_normal_inactive.png"
+theme.titlebar_maximized_button_focus_inactive  = "@AWESOME_THEMES_PATH@/default/titlebar/maximized_focus_inactive.png"
+theme.titlebar_maximized_button_normal_active = "@AWESOME_THEMES_PATH@/default/titlebar/maximized_normal_active.png"
+theme.titlebar_maximized_button_focus_active  = "@AWESOME_THEMES_PATH@/default/titlebar/maximized_focus_active.png"
+
+theme.wallpaper = "@AWESOME_THEMES_PATH@/default/background.png"
+
+-- You can use your own layout icons like this:
```
Comment (actionless) [inline themes/xresources/theme.lua.in]: Also it's possible to check here if background is darker than #888888 and if not then don't add `w` in the end of layout filenames

File: themes/xresources/theme.lua.in [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+end
+
+
+--- Compute resulting fontsize applying dpi value from xrdb.
+-- @tparam size int Font size
+function xresources.compute_fontsize(size)
+  if not xresources.dpi then
+    local output = io.popen("xrdb -query")
+    local query = output:read('*a')
+    xresources.dpi = tonumber(string.match(query, "dpi:[%s]+([%d]+)"))
```
Comment (blueyed) [inline themes/xresources/theme.lua.in]: I am getting `Xft.dpi:	96` from `xrdb -query`, although I do not set it explicitly.
It should be 125 instead.
Where does this come from?
(see also http://www.thinkwiki.org/wiki/Installing_Gentoo_on_a_ThinkPad_X220#DPI)

File: themes/xresources/theme.lua.in (hunk 2 above)
Comment (blueyed) [inline themes/xresources/theme.lua.in]: Good idea!

File: themes/xresources/theme.lua.in (hunk 3 above)
Comment (actionless) [inline themes/xresources/theme.lua.in]: 96 is the default value

Comment (blueyed) [inline themes/xresources/theme.lua.in]: Ah. Maybe `xdpyinfo` could be used then instead?

This setting can (presumably) also be different per screen, could this be handled?

Comment (actionless) [inline themes/xresources/theme.lua.in]: GTK and Qt have the same behavior -- it seems like fontsize depends on the value from xrdb (not xdpyinfo); also it's the same on both displays.

Comment (actionless) [inline themes/xresources/theme.lua.in]: mb it can be grabbed with `lgi`?

Comment (blueyed) [inline themes/xresources/theme.lua.in]: > mb current dpi value used in GTK can be grabbed using lgi?

@Elv13 might know this..

**Comment by blueyed** [comment]
Do you think setting `dpi` on the theme, and/or exporting the `xresources.compute_fontsize` function makes sense?
I'd like to use it for setting `font` myself, after `beautiful.init`.

I think the `dpi` helpers might also get added to `beautiful` itself, making it available for all themes.

**Comment by actionless** [comment]
I think doing that in the theme is ok, because if I would be done on some global level it can be a breaking change.

Exporting can be useful. Where should i put it, as `beautiful.xresources.compute_fontsize` and `beautiful.xresources.get_current_theme` or somehow else?

File: themes/xresources/theme.lua.in [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+-- Returns a table with keys 'bg', 'fg' and '0'..'15'
+function xresources.get_current_theme()
+  local colors = {}
+  local output = io.popen("xrdb -query")
+  local query = output:read('*a')
+  output:close()
+  for i,color in string.gmatch(query, "%*color(%d+):[^#]*(#[%a%d]+)") do
+    colors[i] = color
+  end
+  if not colors["15"] then
```
Comment (actionless) [inline themes/xresources/theme.lua.in]: mb show notification what "xrdb is missing or some of the colors are not set"?

Comment (blueyed) [inline themes/xresources/theme.lua.in]: Sounds good, as a naughty notification, I suppose?
Or maybe rather as a warning printed to stdout (for the log)?

**Comment by blueyed** [comment]
> `beautiful.xresources.compute_fontsize` and `beautiful.xresources.get_current_theme`

Sounds good.

**Comment by blueyed** [comment]
I wonder how using `XSETTINGS` is related in this context: could `xsettingsd` be queried for something in this regard, too?
Or is it unrelated, because it would only have settings for the gtk theme being used etc?

**Comment by actionless** [comment]
>> could xsettingsd be queried for something in this regard, too?

`dump_xsettings` is analogue to `xrdb -query`. I think the latter can be called as a fallback if `dump_xsettingsd` won't return any dpi value.

Also i just noticed what in xsettings GTK ColorPalette is exposed so that opens an opportunity to create one more theme which will follows GTK theme instead of xrdb :-)

**Comment by actionless** [comment]
@blueyed i hope i fixed all the comments

**Comment by actionless** [comment]
or mb separate `beautiful.lua` to `beautiful/init.lua` and `beautiful/xresources.lua`?

**Comment by actionless** [comment]
and mb also generate using cairo the same wallpaper (https://raw.githubusercontent.com/awesomeWM/awesome/master/themes/zenburn/zenburn-background.png) but using current colors?

**Comment by blueyed** [comment]
> Mb it will be good to separate beautiful.lua to beautiful/init.lua and beautiful/xresources.lua?

:+1: 

> And also what do you think about generating with cairo the same wallpaper (https://raw.githubusercontent.com/awesomeWM/awesome/master/themes/zenburn/zenburn-background.png) but using current colors?

That would be awesome, of course! :)

**Comment by actionless** [comment]
added wallpaper: http://i.imgur.com/d8GwbQF.png

**Comment by actionless** [comment]
@blueyed i hope i finished with it

**Comment by blueyed** [comment]
Great!

I'm moving it to the "next" milestone, and will test it some more during the next days.

**Comment by blueyed** [comment]
@actionless 
I think this could be merged into a single commit.

Do you want to do it?
Otherwise I can do it when merging it.

**Comment by actionless** [comment]
@blueyed squashed

**Comment by blueyed** [comment]
@actionless 
Thanks, although the commit message should now describe this single commit, and not (necessarily / only) the steps to get there (the previous commit messages).
Please fix this.

While at it, you can add `Closes https://github.com/awesomeWM/awesome/pull/229`, and it will automatically close the PR when the commit is cherry-picked / pushed to master.

**Comment by actionless** [comment]
fixed

**Comment by actionless** [comment]
>which should take into account the dpi setting
not working for me, it seems like it's always using the default value (96)

**Comment by actionless** [comment]
to rephrase one of my comments on commit:
where it's better to retrieve from xserver itself (like in i3) or from xrdb (like in GTK, Qt, and recently in firefox and chrome)

**Comment by actionless** [comment]
i did not squahed newer commits for more observable changeset

**Comment by actionless** [comment]
@blueyed 
@psychon 
anything else need to be done?

**Comment by blueyed** [comment]
I've found that 37ecc3c2cff6ddcffc0a08c7df3da1e38de65078 causes my fonts to be bigger.

I am using a DPI of 125, that is set via `xrdb --dpi 125`.

I'm not sure that this is right, and might indicate that I find the defaults too big?! (which was compensated by my wrong DPI setting) :)

Also, the moving from beautiful.lua to a module has some issues for the typical "make install" (https://github.com/awesomeWM/awesome/issues/244).  You need to remove / cleanup "build" and the installed file system, when switching.

Given that my "fonts are now all bigger" is intentional, I'm good to merge this.

**Comment by actionless** [comment]
and what about retrieving dpi settings from xserver via xcb, should i add that?

**Comment by actionless** [comment]
i did not understood from the previous comments, should i take into account xserver dpi (the one which can be set in xorg.conf or via `xrandr --dpi`)?

and if so then with each "priority"?

**Comment by blueyed** [comment]
It seems like that patch makes it better: the fonts appear to be OK, but the
height and icon size of e.g. the taskbar gets upped.

0 uses the DPI of the (previous?) font context.

I'm not sure about this still.

    diff --git i/lib/wibox/widget/textbox.lua.in w/lib/wibox/widget/textbox.lua.in
    index 499c4d4..f55aae7 100644
    --- i/lib/wibox/widget/textbox.lua.in
    +++ w/lib/wibox/widget/textbox.lua.in
    @@ -132,7 +132,7 @@ local function new(text, ignore_markup)
         end
     
         local ctx = PangoCairo.font_map_get_default():create_context()
    -    ctx:set_resolution(beautiful.xresources.get_dpi())
    +    ctx:set_resolution(0)
         ret._layout = Pango.Layout.new(ctx)
     
         ret:set_ellipsize("end")

**Comment by blueyed** [comment]
> i did not understood clearly from the previous comments, should i take into account xserver dpi (the one which can be set in xorg.conf or via xrandr --dpi)?

I cannot tell.  I am setting my DPI via `xrandr --dpi`.
I don't know what Pango or awesome use and pick up by default.

**Comment by actionless** [comment]
i got even more confused, let's clarify that on IRC then

**Comment by blueyed** [comment]
It's maybe better to keep it here.

Probably ignore just my last two comments: I think it's mainly about me having an incorrect DPI setting before (changed from 96 to 125).

For me it seems like the fonts being bigger is what I see from other apps (which is expected), but then the change in awesome (esp. the tasklist and taglist) seemed to be too much.  But then I was using e.g. "Ubuntu Sans 10" for the tasklist especially.

But I'd like to get this verified / tested by someone else before merging.

Also, the splitting of `beautiful` into a module would be nice to get ACKed by @psychon.

**Comment by actionless** [comment]
i have learned how to read xrdb values using `X11/Xresource.h`

should i rewrite that part?

**Comment by Elv13** [comment]
If it avoid the `popen()`, then yes, that would be appreciated. `popen()` cause an I/O interrupt and you never know what can go wrong with those.

**Comment by actionless** [comment]
@Elv13 what to do with `xsettingsd` then?
1) allow popen there
2) drop it's support
3) investigate connecting to settings daemon from c code

**Comment by Elv13** [comment]
`dump_xsettings` isn't even a valid command on my system. So if it can be in another PR, then great. I just want to be clear, I don't like `popen`, but that isn't a blocker for me. As Awesome is single threaded, each `popen` block the main thread _and_ all input to clients. Even a few milliseconds lost due to I/O scheduling in the kernel will make the apps feel less snappy. I know your code isn't called that much after initialization, but this is my opinion on `popen` in general. For longer commands, just as fetching weather from the network like some config has, LGI provide access to both GIO async file access and async sockets (including local commands). This doesn't really apply here as the commands need to be synchronous.

**Comment by actionless** [comment]
ok, i'll drop it and rewrite it as c module.
but i am not sure create a new one (for one function?) or to some of the existing ones?

offtopic: on my config awesome crashes when using lgi calling async open from other async open's callback but i did not digged deeper to have enough info for a bug report

**Comment by actionless** [comment]
@Elv13 
>Edit: Unless it cause a new dependency on xlib, that would be bad

isn't `X11/Xresource.h` is provided by 
https://github.com/awesomeWM/awesome/blob/master/awesomeConfig.cmake#L134 ?

**Comment by actionless** [comment]
git hanged on travis

File: xrdb.c [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+ * \lparam string xrdb name, ie "background" or "color0"
+ * \lreturn string xrdb value or nil if not exists. \
+ */
+int luaA_xrdb_get_value(lua_State *L) {
+  if (!xrmdb)
+    xrdb_init();
+
+  char *resource_type;
+  int resource_code;
+  XrmValue resource_value;
```
Comment (actionless) [inline xrdb.c]: should i free it after somehow?

**Comment by actionless** [comment]
re: travis

mb use some XCB tarball instead of checking it out from git? i have noticed what it fails on that step pretty often (not only in this PR)

File: xrdb.c [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+Display *dpy;
+XrmDatabase xrmdb;
+
+/* \brief open X display and X Resources DB
+ */
+static void xrdb_init(void) {
+  /**/
+  /* Open display with Xlib */
+  /**/
+  char *displayname = NULL;
```
Comment (actionless) [inline xrdb.c]: if it's NULL it will be retrieved automatically but maybe somehow extract it from xcb connection in globalconf?

**Comment by blueyed** [comment]
> re: travis
> 
> mb use some XCB tarball instead of checking it out from git? i have noticed what it fails on that step pretty often (not only in this PR)

Yes, I'll improve this (https://github.com/awesomeWM/awesome/compare/master...blueyed:travis-wget-instead-of-git-clone).  But it seems to often hang at "Scanning dependencies of target check", too?!

**Comment by ff2000** [comment]
Daniel Hahler wrote:
> > re: travis
> > 
> > mb use some XCB tarball instead of checking it out from git? i have noticed what it fails on that step pretty often (not only in this PR)
> 
> Yes, I'll improve this (https://github.com/awesomeWM/awesome/compare/master...blueyed:travis-wget-instead-of-git-clone).  But it seems to often hang at "Scanning dependencies of target check", too?!

You may want to use "ninja" cmake generator instead of Gnu Makefile. I had the same issue when building webkit-gtk (30 minutes of the complete build time was spent in "scanning dependencies of target...") and switching to ninja solved the issue.

**Comment by blueyed** [comment]
@ff2000
Thanks for the suggestion, but actually the tests are hanging (I had edited my comment above).

**Comment by actionless** [comment]
so what else should be done from my side?

**Comment by blueyed** [comment]
@actionless 
There were some remaining questions from you (which I could not answer).
Is everything clear / ready from your side?

**Comment by actionless** [comment]
i am not sure about the c part, but it's only 60 lines to check

**Comment by blueyed** [comment]
> i am not sure about the c part, but it's only about 60 lines to check

@psychon 
Can you please review this?

Apart from that, I would like to get feedback from others in general - i.e. regarding the fixed DPI adjustments, where fonts would appear larger/smaller than before.

**Comment by ff2000** [comment]
I just tried the patch, checked xresources and zenburn themes. Fonts appear in the same size as without the patch. The default font always was too small for me. I can't tell if it's supposed to be that way. Any hints?
At least it isn't worse than before ;)

**Comment by actionless** [comment]
@ff2000 
set `Xft/DPI` in ~/.Xresources and run `xrdb -merge ~/.Xresources`:
https://wiki.archlinux.org/index.php/X_resources#Xft_resources

**Comment by ff2000** [comment]
Thx :)
I actually had to look up how to get my DPI. It was 118 instead of 96, which X seems to default to. Now everything looks really nice (and readable)!

There is an issue with the colors. If I specify them directly in my .Xresources everything works. But I like to change color schemes from time to time (I am red/green blind and I often struggle with readability of highlighted source code, so I change themes when I see a screenshot of a potentially better working one - getting WORKING themes work both urxvt and vim seems to be quite an issue). To make it easier I store the colors in a seperate file under ~/.xcolors and #include it in my .Xresources. xresources theme misses those colors.
kde also has issues with this approach - its xrdb-module parses .Xresources file without following #include'd files and completely messes up my urxvt as it actually resets xrdb :(

**Comment by actionless** [comment]
@ff2000 sorry, i did not understood what is the issue

in my config colors are included too (even folder name is the same :) ) and it works

**Comment by blueyed** [comment]
@ff2000 
Thanks for your feedback!
As @actionless said, the colors get queried through `xrdb`, not by parsing the `.Xresources` file, so that should work.  I am setting them via `#include`, too.

(I am using the Solarized theme btw, and switch between the light/dark variants - it works great with URXvt and Vim (using the adjusted 16 color palette) once setup).

**Comment by ff2000** [comment]
Error found - on my side ;)
The theme I use prefixed the colors with URxvt, as it's meant to be an urxvt-theme ;)

@blueyed I currently use https://github.com/junegunn/seoul256.vim. I used solarized before but I had issues in vim. Don't know if it was the highlighting or the colorscheme, many colors were too similar/the same. I will try your urxvt/vim setup (hope your dotfiles aren't outdated ;)), hope it won't break output of "eix"...

**Comment by blueyed** [comment]
@ff2000 
My dotfiles should be fairly up-to-date (in this regard).
I am using this theme for Vim: https://github.com/blueyed/vim-colors-solarized
The important option is `let g:solarized_termcolors=16`.

File: xrdb.c [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+                   // needed
+  if (!(xrmdb = XrmGetDatabase(dpy))) {
+
+    /* taken from xpbiff: */
+    /* >> what a hack; need to initialize dpy->db */
+    (void)XGetDefault(dpy, "", "");
+    /**/
+
+    if (!(xrmdb = XrmGetDatabase(dpy)))
+      fatal("Can't open xrdb\n");
```
Comment (blueyed) [inline xrdb.c]: Would these calls to `fatal` prevent startup of awesome?
If so, maybe `warn` should be used instead? (not that it's likely to happen, but still).

File: xrdb.c (hunk 6 above)
Comment (blueyed) [inline xrdb.c]: Yes, I also think that `globalconf.connection` should/could be used here somehow (also below).

Comment (blueyed) [inline xrdb.c]: http://www.x.org/wiki/guide/xlib-and-xcb/#index2h2 says
> As mentioned above, XCB provides a method for incremental conversion from Xlib to XCB. One can use libX11 to open the display and pass the Display pointer it returns to existing code, toolkits, and libraries. To call an XCB function, one can convert the Display pointer to an xcb_connection_t pointer for the same connection. This enables calling into Xlib and XCB from the same application.

Not sure if that works also in the other direction?!

File: xrdb.c [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+#include <X11/Xresource.h>
+
+#include <string.h>
+
+Display *dpy;
+XrmDatabase xrmdb;
+
+/* \brief open X display and X Resources DB
+ */
+static void xrdb_init(void) {
```
Comment (actionless) [inline xrdb.c]: seems like it shouldn't be static?

**Comment by psychon** [comment]
Look at e.g. commit aaa771f8879e0a16dfcaa7f5f6c60cc356498c18 on how to mix Xlib with libxcb. You'll want to re-add the call to `XkbIgnoreExtension` (although I guess it isn't needed anymore, but still) and the code around that. Also, don't use global variables, but store things in `globalconf`. Dunno about the rest of this.

**Comment by actionless** [comment]
@psychon where should i init xlib, in awesome.c (there it was before https://github.com/awesomeWM/awesome/commit/aaa771f8879e0a16dfcaa7f5f6c60cc356498c18) or in xrdb.c (there it is now)?

**Comment by actionless** [comment]
* rebased to resolve merge conflict

**Comment by psychon** [comment]
@actionless To avoid a ping-pong game, here is what I have in sense: https://gist.github.com/psychon/13189654af8cafb3838b

**Comment by actionless** [comment]
@psychon thanks for the patch :-)
i also cherry-picked https://github.com/awesomeWM/awesome/pull/271 to fix awesome start

**Comment by blueyed** [comment]
This needs to be rebased on master, for https://github.com/awesomeWM/awesome/commit/32b584258bb2fc3dfc744071a3b73461bb64eacf.  After that, it seems to work good!

btw: for what it's worth, it required a clean build (after being on another
branch (cmake-no-recurse-into-build-dir)):

    Linking C executable awesome
    /usr/bin/ld: CMakeFiles/awesome.dir/awesome.c.o: undefined reference to symbol 'XSetEventQueueOwner'
    /usr/lib/x86_64-linux-gnu/libX11-xcb.so.1: error adding symbols: DSO missing from command line
    collect2: error: ld returned 1 exit status

I wonder if that's related to that we should define/set the input files explicitly?!

**Comment by actionless** [comment]
@blueyed rebased

i did not squashed the commits in case if someone would like to review the latest changes more detailed

**Comment by blueyed** [comment]
I've gone ahead and merged it finally - in a single commit, adjusting your/the commit message.

Thanks a lot again for this!

**Pull request closed without merging**